### PR TITLE
Adds _std/get_average_color, uses it on custom chef recipes

### PR DESCRIPTION
**Pull request by TDHooligan**
## Why's this needed? 
Fixes coloration of ingredients in custom foods, so they aren't bright red like this fried cheese cake:
image

Custom foods will now look like so:
image
Slurry pie cake, fried cheese cake, fried meat cake

## Changelog 

```changelog
(u)TDHooligan
(+)X Cakes and similar recipes will now more closely match the color of their ingredients, rather than being bright red.
```

### CODE REVIEW

File: _std/color.dm [hunk 1]
```diff
@@ -274,6 +274,29 @@ proc/hsv_transform_color_matrix(h=0.0, s=1.0, v=1.0)
...
+			if (!pixColor)
+				continue
+			var/rgba = rgb2num(pixColor)
+			var/weight = length(rgba) >= 4 ? rgba[4] / 255 : 1
+			total += weight
+			rSum += rgba[1] * weight
+			gSum += rgba[2] * weight
+			bSum += rgba[3] * weight
+	if (total == 0)
+		return "#FF0000"
```
Comment (pali6) [inline _std/color.dm:297]: I'd probably either make this return `null` or `#00000000` (i.e. fully transparent) and then handle the turning to red where this is called.

File: _std/color.dm [hunk 2]
```diff
@@ -274,6 +274,29 @@ proc/hsv_transform_color_matrix(h=0.0, s=1.0, v=1.0)
 		0, 0, 0, 0
 	)
 
+proc/get_average_color(icon/I, xPixelInterval = 4, yPixelInterval = 4)
```
Comment (pali6) [inline _std/color.dm:277]: I think the default should be 1 and not 4 here.

File: _std/color.dm [hunk 3]
```diff
@@ -274,6 +274,29 @@ proc/hsv_transform_color_matrix(h=0.0, s=1.0, v=1.0)
...
+	var/rSum  = 0
+	var/gSum  = 0
+	var/bSum  = 0
+	var/total = 0
+	var/xReps = round(32/xPixelInterval)
+	var/yReps = round(32/yPixelInterval)
```
Comment (pali6) [inline _std/color.dm:283]: Check the icon's size using the byond proc, don't hardcode 32.

File: code/modules/chemistry/tools/food_and_drink.dm [hunk 4]
```diff
@@ -32,6 +32,13 @@ ABSTRACT_TYPE(/obj/item/reagent_containers/food)
 				return 1
 		return 0
 
+	proc/get_food_color()
+		if (food_color) // keep manually defined food colors
+			return food_color
+		var/icon/I = istype(src.icon, /icon) ? src.icon : icon(src.icon, src.icon_state)
+		food_color = get_average_color(I)
+		return food_color
```
Comment (pali6) [inline code/modules/chemistry/tools/food_and_drink.dm:40]: Hmm, this caches the color which is probably good as it results in skipping the get_average_color call in future calls on this objects. However, note that this also means that food_color won't get updated if this changes icon_state. Though that's probably unlikely to matter much.

I'd keep it like you wrote it, this comment is more of a sidenote.

**Review by pali6 — CHANGES_REQUESTED** [review]

File: _std/color.dm (hunk 2 above)
Comment (TDHooligan) [reply _std/color.dm:277]: doing 1000 checks seems a bit intense for general usage, doesn't it?

**Comment by pali6** [comment]
Actually instead of using the `xRep` stuff etc. you should use the `for(var/x in 1 to icon_width step xPixelInterval)` syntax (I hope I remember it correctly).

File: code/modules/chemistry/tools/food_and_drink.dm (hunk 4 above)
Comment (TDHooligan) [reply code/modules/chemistry/tools/food_and_drink.dm:40]: yeah, ultimately some of the food procs check the color several times and it seemed more apt to just cache it than rebuild it.

File: code/modules/chemistry/tools/food_and_drink.dm [hunk 5]
```diff
@@ -9,7 +9,7 @@ ABSTRACT_TYPE(/obj/item/reagent_containers/food)
 	var/heal_amt = 0
 	var/needfork = 0
 	var/needspoon = 0
-	var/food_color = "#FF0000" //Color for various food items
+	var/food_color = null //Color for various food items
```
Comment (ZeWaka) [inline code/modules/chemistry/tools/food_and_drink.dm:12]:
```suggestion
	/// Color for various food items
	var/food_color = null 
```

**Review by pali6 — APPROVED** [review]

**Pull request merged**